airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,214 @@
+      serialized DAGs are updated in DB when a file gets processed by scheduler,
+      to reduce DB write rate, there is a minimal interval of updating serialized DAGs.
+    * ``[scheduler] dag_dir_list_interval = 300`` (s):
+      interval of deleting serialized DAGs in DB when the files are deleted, suggest
+      to use a smaller interval such as 60
+
+    It is used by webserver to load dagbags when ``store_serialized_dags=True``.
+    Because reading from database is lightweight compared to importing from files,
+    it solves the webserver scalability issue.
+    """
+    __tablename__ = 'serialized_dag'
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    fileloc = Column(String(2000), nullable=False)
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer, nullable=False)
+    data = Column(JSON, nullable=False)
+    last_updated = Column(UtcDateTime, nullable=False)
+
+    __table_args__ = (
+        Index('idx_fileloc_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, dag: 'DAG'):
+        from airflow.serialization import SerializedDAG  # noqa # pylint: disable=redefined-outer-name
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = self.dag_fileloc_hash(self.fileloc)
+        self.data = SerializedDAG.to_dict(dag)
+        self.last_updated = timezone.utcnow()
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        return int.from_bytes(
+            hashlib.sha1(full_filepath.encode('utf-8')).digest()[-2:], byteorder='big', signed=False)
+
+    @classmethod
+    @db.provide_session
+    def write_dag(cls, dag: 'DAG', min_update_interval: Optional[int] = None, session=None):
+        """Serializes a DAG and writes it into database.
+
+        :param dag: a DAG to be written into database
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        :param session: ORM Session
+        """
+        log.debug("Writing DAG: %s to the DB", dag)
+        # Checks if (Current Time - Time when the DAG was written to DB) < min_update_interval
+        # If Yes, does nothing
+        # If No or the DAG does not exists, updates / writes Serialized DAG to DB
+        if min_update_interval is not None:
+            if session.query(exists().where(
+                and_(cls.dag_id == dag.dag_id,
+                     (timezone.utcnow() - timedelta(seconds=min_update_interval)) < cls.last_updated))
+            ).scalar():
+                return
+        session.merge(cls(dag))
```

Comment:
Makes more sense to move the `log.debug("Writing DAG: %s to the DB", dag)` line here, because the interval check is done before writing the DAG.

Reply:
👍